homeassistant/components/modbus/__init__.py

```diff
@@ -199,6 +211,20 @@
             vol.Optional(CONF_MIN_TEMP, default=5): cv.positive_int,
             vol.Optional(CONF_STEP, default=0.5): vol.Coerce(float),
             vol.Optional(CONF_TEMPERATURE_UNIT, default=DEFAULT_TEMP_UNIT): cv.string,
+            vol.Optional(CONF_HVAC_ACTION, default=0): cv.positive_int,
```

Comment:
The user should not choose HVAC mode, that is normally something the program does (or more correctly the automation connected).

Some climate controls deliver the hvac mode, but there are no standard in modbus for that.

Reply:
That is correct - there is no "standard" for the hvac mode. I would argue the same is true for a modbus "climate" device in general, this entire integration doesn't really follow a standard. It's supposed to be generic enough that with enough configuration of registers, types, etc, that Home Assistant allows the device to be used as a thermostat. I want to add support for more thermostat features - as generic as possible.

That being said - I spent quite a bit of time writing this to be as generic as possible, and to allow the user to map a device register to either the active mode `HVAC_ACTION` and/or supported modes `HVAC_MODE`. The user then specifies when Home Assistant reads those two registers, what the values actually mean in the context of a thermostat (the values configuration variables).

The active hvac state is read out of the device and interpreted according to the `hvac_action_supported` list.
The hvac mode can either be read from the device and interpreted according to the `hvac_mode_supported` list, or it can be written to the device using the `async_set_hvac_mode()` function.

The only possible thing that I can think to do to make this more generic than it already is would be to allow the user to set the register types (holding register, etc), count (number of bytes), structure (python struct type), etc. It would add more configuration variables, but I'm willing to do it if that's what's necessary to get this PR approved.